pkg/operator/metrics/collector.go

```diff
@@ -0,0 +1,215 @@
+package metrics
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+)
+
+const (
+	openshiftOperatorNamePrefix = "operator_v1_openshift_io"
+	workkey                     = "key"
+)
+
+// OperatorStateGetter knows how to get the current operator state and provides operator client informer.
+type OperatorStateGetter interface {
+	GetOperatorState() (spec *operatorv1.OperatorSpec, status *operatorv1.OperatorStatus, staticPodStatus *operatorv1.StaticPodOperatorStatus, resourceVersion string, err error)
+	Informer() cache.SharedIndexInformer
+}
+
+// operatorMetricUpdater knows how to update a single tracked metric.
+type operatorMetricUpdater interface {
+	Update(*operatorv1.OperatorSpec, *operatorv1.OperatorStatus, *operatorv1.StaticPodOperatorStatus) error
+}
+
+type operatorMetricUpdateFn func(*operatorv1.OperatorSpec, *operatorv1.OperatorStatus, *operatorv1.StaticPodOperatorStatus) ([]metricValue, error)
+
+type Collector struct {
+	operatorClient  OperatorStateGetter
+	operatorMetrics []operatorMetricUpdater
+	targetNamespace string
+
+	queue          workqueue.RateLimitingInterface
+	hasSynced      cache.InformerSynced
+	collectorMutex sync.RWMutex
+}
+
+// metricValue represents a single metric value sample
+type metricValue struct {
+	value  float64
+	labels []string
+}
+
+// operatorMetrics represents a single metric definition
+type operatorMetric struct {
+	prometheusDesc *prometheus.Desc
+	valueType      prometheus.ValueType
+
+	// values are the recently observed metric values provided by the updateFn
+	values []metricValue
+
+	// updateFn know how to gather the values for this metric
+	updateFn operatorMetricUpdateFn
+
+	// valueMutex protects from values changes
+	valueMutex sync.RWMutex
```

Comment:
just `values`? By convention, but locks directly above (without empty lines) the objects they protect.